client/network/sync/src/engine.rs

```diff
@@ -524,13 +619,194 @@ where
+			Poll::Ready(Some(pre_validation)) => {
+				self.deallocate_slot_for_block_announce_validation(pre_validation.who());
+
+				let res = match pre_validation {
+					PreValidateBlockAnnounce::Process { is_new_best, who, announce } => {
+						trace!(
+							target: "sync",
+							"Finished block announce validation: from {:?}: {:?}. local_best={}",
+							who,
+							announce.summary(),
+							is_new_best,
+						);
+						self.chain_sync.on_validated_block_announce(is_new_best, who, &announce);
+						PollBlockAnnounceValidation::Nothing { is_new_best, who, announce }
+					},
+					PreValidateBlockAnnounce::Failure { who, disconnect } => {
+						debug!(
+							target: "sync",
+							"Failed announce validation: {:?}, disconnect: {}",
+							who,
+							disconnect,
+						);
+						PollBlockAnnounceValidation::Failure { who, disconnect }
+					},
+					PreValidateBlockAnnounce::Error { who } => {
+						debug!(
+							target: "sync",
+							"Ignored announce validation from {:?} due to internal validation error",
+							who,
+						);
+						PollBlockAnnounceValidation::Skip
+					},
+				};
+
+				Poll::Ready(res)
+			},
+			_ => Poll::Pending,
+		}
+	}
+
+	/// Checks if there is a slot for a block announce validation.
+	///
+	/// The total number and the number per peer of concurrent block announce validations
+	/// is capped.
+	///
+	/// Returns [`AllocateSlotForBlockAnnounceValidation`] to inform about the result.
+	///
+	/// # Note
+	///
+	/// It is *required* to call [`Self::deallocate_slot_for_block_announce_validation`] when the
+	/// validation is finished to clear the slot.
+	fn allocate_slot_for_block_announce_validation(
+		&mut self,
+		peer: &PeerId,
+	) -> AllocateSlotForBlockAnnounceValidation {
+		if self.block_announce_validation.len() >= MAX_CONCURRENT_BLOCK_ANNOUNCE_VALIDATIONS {
+			return AllocateSlotForBlockAnnounceValidation::TotalMaximumSlotsReached
+		}
+
+		match self.block_announce_validation_per_peer_stats.entry(*peer) {
+			Entry::Vacant(entry) => {
+				entry.insert(1);
+				AllocateSlotForBlockAnnounceValidation::Allocated
+			},
+			Entry::Occupied(mut entry) => {
+				if *entry.get() < MAX_CONCURRENT_BLOCK_ANNOUNCE_VALIDATIONS_PER_PEER {
+					*entry.get_mut() += 1;
+					AllocateSlotForBlockAnnounceValidation::Allocated
+				} else {
+					AllocateSlotForBlockAnnounceValidation::MaximumPeerSlotsReached
+				}
+			},
+		}
+	}
+
+	/// Should be called when a block announce validation is finished, to update the slots
+	/// of the peer that send the block announce.
+	fn deallocate_slot_for_block_announce_validation(&mut self, peer: &PeerId) {
+		match self.block_announce_validation_per_peer_stats.entry(*peer) {
+			Entry::Vacant(_) => {
+				error!(
+					target: "sync",
+					"💔 Block announcement validation from peer {} finished for that no slot was allocated!",
+					peer,
+				);
+			},
+			Entry::Occupied(mut entry) => {
+				*entry.get_mut() = entry.get().saturating_sub(1);
```

Comment:
This could be `checked_sub()` with a print if something's wrong

Reply:
Done